Stop hidden products from being linked in cart and checkout blocks

opr opened the pull request:
Fixes #3370 

### Screenshots

<img width="339" alt="Screenshot 2020-11-17 at 16 30 11" src="https://user-images.githubusercontent.com/5656702/99417840-4af99780-28f2-11eb-991e-6f4c7f890b77.png">

<img width="528" alt="Screenshot 2020-11-17 at 16 32 14" src="https://user-images.githubusercontent.com/5656702/99418009-75e3eb80-28f2-11eb-83ad-cd894e6ee31f.png">

The "T-shirt with Logo" product is hidden from the catalogue, and therefore its name or image does not link to the product page.
### How to test the changes in this Pull Request:

1. Create a page that contains the checkout and cart blocks.
2. Find a product in the dashboard and set its catalogue visibility to hidden.
3. Add that product to your cart.
4. Visit the page with the checkout and cart blocks and ensure the name of the hidden product does not link to the product page in the checkout sidebar and in the line items in the cart.
5. Repeat steps 2-4 with the catalogue visibility set to "Search results only".

opr committed:
Add catalog_visibility to CartItemSchema.php

This is used to get whether the product is visible in the catalogue, visible in the shop only, visible in search results only, or visible everywhere. We need to know this so we can pass it to the ProductImage and ProductName components.

opr committed:
Remove links from CartLineItemRow if not visible in catalogue

Added catalog_visibility to lineItems prop, and when the product is not visible in the catalogue do not wrap product image in a link. Also pass down the hasLink prop to ProductName.

opr committed:
Add hasLink prop to ProductName

When this prop is false we should not output the link around the product name. This is for when the product is hidden from the catalogue but we still want to show its name somewhere.

opr committed:
Add tests and storybook for ProductName

opr committed:
Add catalog_visibility check to OrderSummaryItem

When the catalogue visibility of a product is set to hidden or search, then the product name in the checkout sidebar should not be hyperlinked.

budzanowski commented on assets/js/base/components/cart-checkout/product-name/index.js:
```diff
@@ -9,8 +9,13 @@ import { decodeEntities } from '@wordpress/html-entities';
+} ) => {
+	return hasLink ? (
 		// we use tabIndex -1 to prevent the link from being focused, pointer-events
 		// disabled click events, so we get an almost disabled link.
```
Would using `disabled` not be enough?

opr replied:
It seems `disabled` doesn't stop people clicking on the link, it stops people _tabbing_ to the link but not clicking it. I notice the comment 

```
// we use tabIndex -1 to prevent the link from being focused, pointer-events
// disabled click events, so we get an almost disabled link.
```

but notice also that `pointer-events: none` never gets applied to the `a` here. 

My counter question is: why do we keep the `a` here if it's supposed to be disabled?

I propose that we use the `disabled` prop, so no need to add `hasLink` but keep the logic that outputs a `span` if the link is disabled, or outputs an `a` if it is not.

budzanowski commented on assets/js/base/components/cart-checkout/product-name/index.js:
```diff
@@ -9,8 +9,13 @@ import { decodeEntities } from '@wordpress/html-entities';
+	name,
+	permalink,
+	disabled = false,
+	hasLink = true,
```
ProductName.propTypes need an update after adding this.

opr replied:
Won't be needed since we are now using `disabled`.

budzanowski commented:
Left two small comments but it is good otherwise. LGTM.

opr committed:
Reverse logic for hiding link on product image & disabling link on name

Following a point from @budzanowski we do not need the hasLink prop, making use of disabled is probably a better idea.

opr committed:
Remove tabindex from a in ProductName & output span if name is disabled

This change removes the need to pass a tabindex to the a in ProductName. This is because a disabled ProductName will now never output an a tag. When the ProductName is disabled a span is output instead, which has no tabindex by default.

This change also reverses the logic to decide whether the a or span should be output so as to make the code more readable and flow better.

opr committed:
Update storybook and tests/snapshots for ProductName
